### DISCUSSION TITLE
How to use `CommentProvider` if language was create using `createServicesForGrammar`?

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1453] [Q&A] [answered]
Hi there,

I'm currently trying to create a sample grammar for a test case that has JSDoc using `createServicesForGrammar`:

```ts
it('should handle grammar with one greedy rule', async () => {
  const grammar = `
      grammar Test
      /** Rule */
      entry Rule: 'rule' num=INT;
      /** INT */
      terminal INT: /\\d+/;
    `;
    const services = await createServicesForGrammar({ grammar });
    services.Grammar.rules.forEach((rule) => {
      expect(services.documentation.CommentProvider.getComment(rule)).toBeDefined();
    });
});
```

The above code fails because the `CommentProvider#getComment` always returns `undefined`.

Is this expected? It's working fine if I created the language using `createDefaultCoreModule` and similar functions.

**Comment by msujew** [comment]
Hey @Yokozuna59,

you're asking the language specific service for comments on the grammar language. You need to ask the grammar service explicitly if you want to get comments for grammars. Similarly, the language specific comment provider only works for the language in question, i.e. the `Test` language.

Note that the `Test` language has no comment terminal rules, therefore the comment provider will never actually return any comments.

**Reply by Yokozuna59** [reply]
@msujew Could you provide me with a sample code or just show the functions that should solve the issue? I tried several ways to do so, but none of them were correct.

**Reply by msujew** [reply, marked as answer]
You would use the `createLangiumGrammarServices()` method to get the grammar services and with that the grammar comment provider:

```ts
import { createLangiumGrammarServices } from 'langium/grammar';
const grammarServices = createLangiumGrammarServices(EmptyFileSystem);
it('should handle grammar with one greedy rule', async () => {
  const grammar = `
      grammar Test
      /** Rule */
      entry Rule: 'rule' num=INT;
      /** INT */
      terminal INT: /\\d+/;
    `;
    const services = await createServicesForGrammar({ grammar });
    services.Grammar.rules.forEach((rule) => {
      expect(grammarServices.documentation.CommentProvider.getComment(rule)).toBeDefined();
    });
});
```
